leshan-core/src/main/java/org/eclipse/leshan/core/attributes/AttributeSet.java

```diff
@@ -0,0 +1,134 @@
+    
+    public AttributeSet(Attribute...attributes) {
+        if (attributes != null && attributes.length > 0) {
+            for (Attribute attr : attributes) {
+                // Check for duplicates
+                if (attributeMap.containsKey(attr.getName())) {
+                    throw new IllegalArgumentException(String.format("Cannot create attribute set with duplicates (attr: '%s')",
+                            attr.getCoRELinkParam())); 
+                }
+                attributeMap.put(attr.getName(), attr);
+            }
+        }
+    }
+    
+    /**
+     * Creates an attribute set from a list of query params.
+     */
+    public AttributeSet(String[] queryParams) {
+        for (String param : queryParams) {
+            String[] keyAndValue = param.split("=");
+            if (keyAndValue.length != 2) {
+                throw new IllegalArgumentException(String.format("Cannot parse query param '%s'", param));
+            }
+            Attribute attr = Attribute.create(keyAndValue[0], keyAndValue[1]);
+            attributeMap.put(attr.getName(), attr);
+        }
+    }
+
+    public void validate(AssignationLevel assignationLevel) {
+        // Can all attributes be assigned to this level?
+        for (Attribute attr : attributeMap.values()) {
+            if (!attr.canBeAssignedTo(assignationLevel)) {
+                throw new IllegalArgumentException(String.format("Attribute '%s' cannot be assigned to level %s",
+                        attr.getCoRELinkParam(), assignationLevel.name()));
+            }
+        }
+        Attribute pmin = attributeMap.get(AttributeName.MINIMUM_PERIOD);
+        Attribute pmax = attributeMap.get(AttributeName.MAXIMUM_PERIOD);
+        if ((pmin != null) && (pmax != null) && (Long) pmin.getValue() > (Long) pmax.getValue()) {
+            throw new IllegalArgumentException(String.format("Cannot write attributes where '%s' > '%s'",
+                    pmin.getCoRELinkParam(), pmax.getCoRELinkParam()));
+        }
+    }
+
+
+    /**
+     * Returns a new AttributeSet, containing only the attributes that have a matching
+     * Attachment level. 
+     * @param attachment the Attachment level to filter by
+     * @return a new {@link AttributeSet} containing the filtered attributes
+     */
+    public AttributeSet filter(Attachment attachment) {
+        List<Attribute> attrs = new ArrayList<>();
+        for (Attribute attr : getAttributes()) {
+            if (attr.getAttachment() == attachment) {
+                attrs.add(attr);
+            }
+        }
+        return new AttributeSet(attrs.toArray(new Attribute[attrs.size()]));
+    }
+    
+    /**
+     * Merges a AttributeSet onto this instance, overwriting or creating values as necessary.
+     * @param attributes the AttributeSet that should be merged
+     */
+    public void merge(AttributeSet attributes) {
+        for (Attribute sourceAttr: attributes.getAttributes()) {
```

Comment:
Should this return a new Attribute with the merged result? Might be cleaner...

Reply:
you mean a new AttributeSet ?
If we do that we can make AttributeSet immutable.

Reply:
Yes, AttributeSet... :) I'll try to make it immutable, I don't think it will add too much complexity